### DISCUSSION TITLE
Component thermal_expansion_coeff implementation

### DISCUSSION THREAD

**Discussion by chakra34** [idaholab/moose#25548] [Q&A Modules: Solid mechanics] [answered]
Hi,

I am trying to add a code similar to "ComputeThermalExpansionEigenstrain" with something like "ComputeThermalExpansionEigenstrainOrthotropic" where I want to provide 3 values of thermal expansion coefficient. To do that I did the following:

Added a method in RankTwoTensorImplementation.h similar to addIa(a) : 
```c++
void
RankTwoTensorTempl<T>::addDiag(const T & a)
{
  for (auto i : make_range(N))
    (*this)(i, i) += a(i);
}
```

The objective is to have the diagonal components added to the vector (which would be a non-isotropic eigenstrain)

Subsequently, I followed "ComputeThermalExpansionEigenstrain" to write a new code "ComputeThermalExpansionEigenstrainOrthotropic" where I am taking 3 thermal expansion coefficients and the new "computeThermalStrain()" looks something like:

```c++
template <bool is_ad>
ValueAndDerivative<is_ad>
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::computeThermalStrain()
{
  _diagonal_strain_components(1) = _thermal_expansion_coeff_11 * (_temperature[_qp] - _stress_free_temperature[_qp]);
  _diagonal_strain_components(2) = _thermal_expansion_coeff_22 * (_temperature[_qp] - _stress_free_temperature[_qp]);
  _diagonal_strain_components(3) = _thermal_expansion_coeff_33 * (_temperature[_qp] - _stress_free_temperature[_qp]);
  return _diagonal_strain_components ;
}
```

with the _diagonal_strain_components being a std<vector> of size 3 that has the eigenstrain components.

Now my question is-- to add this to the _eigenstrain[_qp] in the computeQpEigenstrain in "ComputeThermalExpansionEigenstrainBase", how do I modify such that when the thermal_strain is a vector it uses the aforementioned .addDiag(thermal_strain.value()) instead of .addIa(thermal_strain.value()) : 

```c++
template <bool is_ad>
void
ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::computeQpEigenstrain()
{
  _eigenstrain[_qp].zero();
  const auto thermal_strain = computeThermalStrain();

  if constexpr (is_ad)
    _eigenstrain[_qp].addIa(thermal_strain);
  else
  {
    _eigenstrain[_qp].addIa(thermal_strain.value());
    (*_deigenstrain_dT)[_qp].zero();
    if (!_use_old_temperature)
      (*_deigenstrain_dT)[_qp].addIa(thermal_strain.derivatives());
  }
}
``` 

One thing I was thinking was to put a flag to check whether the thermal_strain is a scalar or a vector and use accordingly, but I am not sure if that makes sense...
Also, for my case where thermal_strain is a vector can I still use thermal_strain.derivatives() and thermal_strain.values() ?

(P.S. New Moose user and trying my hand at small modifications :( )

Any suggestions/comments would be extremely helpful.

Thanks,
Aritra

**Comment by hugary1995** [comment]
Can't you follow `ComputeCrystalPlasticityThermalEigenstrain` if it doesn't already handle your use case?

**Reply by chakra34** [reply]
Currently, we are going to use it for thermo-elastic problems in small strain, so not crystal plasticity and large strain mechanics. Also with this I want to get me feet wet in coding/understanding MOOSE for some of the future developments we have planned (temperature dependent orthotropic thermal_expansion and elasticity).. so I wanted to implement a relatively simple code.

**Reply by hugary1995** [reply]
I understand what you are trying to do. What I was suggesting is that you can write your code following the `ComputeCrystalPlasticityThermalEigenstrain`. All the operations you need are already done in the source code of that object.

**Reply by chakra34** [reply]
Thanks a lot @hugary1995 for the suggestion, really appreciate it, and I looked at the code and as you mention it has all the requisite operations in the crystal plasticity large strain formulation.

I did manage to write a code that does for simple cubic systems, that seem to work. However, I inherited from ComputeThermalExpansionEigenstrain class. If you get a chance please let me know your thoughts about the approach.

Thanks again,
-Aritra

**Comment by wangzhaohao** [comment]
I think i meet the same problem, my method is use the [`fillFromInputVector`](https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html#a81f42ce0d85006409617f851aceef16d)

**Comment by chakra34** [comment]
@hugary1995 @wangzhaohao  I liked the fillFromInputVector and have used this to accept a vector value for the thermal expansion coeff and then using that to get the eigenstrain. But I think the operator '*' does not work for the following (last function):

```c++

#include "ComputeThermalExpansionEigenstrainOrthotropic.h"

registerMooseObject("TensorMechanicsApp", ComputeThermalExpansionEigenstrainOrthotropic);
registerMooseObject("TensorMechanicsApp", ADComputeThermalExpansionEigenstrainOrthotropic);

template <bool is_ad>
InputParameters
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::validParams()
{
  InputParameters params = ComputeThermalExpansionEigenstrainTempl<is_ad>::validParams();
  params.addClassDescription("Computes eigenstrain due to thermal expansion "
                             "with a 3 coefficients coefficient");
  params.addRequiredParam<std::vector<Real>>(
      "thermal_expansion_coeff_comp", "Thermal expansion coefficient components x y z directions");

  return params;
}

template <bool is_ad>
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::ComputeThermalExpansionEigenstrainOrthotropicTempl(
    const InputParameters & parameters)
  : ComputeThermalExpansionEigenstrainTempl<is_ad>(parameters)
{
  _eigen_ortho_tensor.fillFromInputVector(this->template getParam<std::vector<Real>>("thermal_expansion_coeff_comp"));
}

template <bool is_ad>
ValueAndDerivative<is_ad>
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::computeThermalStrain()
{
  return _eigen_ortho_tensor * (_temperature[_qp] - _stress_free_temperature[_qp]);
}

template class ComputeThermalExpansionEigenstrainOrthotropicTempl<false>;
template class ComputeThermalExpansionEigenstrainOrthotropicTempl<true>;
```
where I want to multiply the _eigen_ortho_tensor (defined as ``` RankTwoTensor ``` in the header) with a scalar which is the difference between the temperatures.

Also, the ```ComputeThermalStrain()``` is used in the ```ComputeThermalExpansionEigenstrainBase``` class as 

```c++
template <bool is_ad>
void
ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::computeQpEigenstrain()
{
  _eigenstrain[_qp].zero();
  const auto thermal_strain = computeThermalStrain();

  if constexpr (is_ad)
    _eigenstrain[_qp].addIa(thermal_strain);
  else
  {
    _eigenstrain[_qp].addIa(thermal_strain.value());
    (*_deigenstrain_dT)[_qp].zero();
    if (!_use_old_temperature)
      (*_deigenstrain_dT)[_qp].addIa(thermal_strain.derivatives());
  }
}
```
This complains when ```computeThermalStrain``` is a RankTwoTensor since it wants a ```ValueAndDerivative```. So what would be the correct datatype to pass through computeThermalStrain() for my case where I am computing the tensor instead of a scalar value?

Thanks again for the help.

-Aritra

**Reply by wangzhaohao** [reply]
i directly pubilc computeEigenstrainBase, i think there should have more simple method.

**Reply by chakra34** [reply]
And then modify computeQPEigenstrain (), yeah..that's an approach. I was thinking along those lines too, but then you would have to get temperature (material property) as well as the stress free temperature at each _qp.

That's why I wanted to inherit from ThermalExpansion class

**Comment by chakra34** [comment]
Following the discussion, I tried modifying ```computeQPEigenstrain()``` function inheriting from ```computeThermalExpansion.C```. So I returned 0.0 for the ```computeThermalStrain ()``` and directly modified the eigenstrain at the [_qp] in my orthotropic class. The code compiles and here's the following source code:


```c++

template <bool is_ad>
InputParameters
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::validParams()
{
  InputParameters params = ComputeThermalExpansionEigenstrainTempl<is_ad>::validParams();
  params.addClassDescription("Computes eigenstrain due to thermal expansion "
                             "with a 3 coefficients coefficient");
  params.addRequiredParam<std::vector<Real>>(
      "thermal_expansion_coeff_comp", "Thermal expansion coefficient components x y z directions");

  return params;
}

template <bool is_ad>
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::ComputeThermalExpansionEigenstrainOrthotropicTempl(
    const InputParameters & parameters)
  : ComputeThermalExpansionEigenstrainTempl<is_ad>(parameters)
{
  _eigen_ortho_tensor.fillFromInputVector(this->template getParam<std::vector<Real>>("thermal_expansion_coeff_comp"));
}

template <bool is_ad>
ValueAndDerivative<is_ad>
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::computeThermalStrain()
{
  return 0.0 ; 
}

template <bool is_ad>
void
ComputeThermalExpansionEigenstrainOrthotropicTempl<is_ad>::computeQpEigenstrain()
{
  // Define Eigenstrain
  _eigenstrain[_qp].zero();
  if constexpr (is_ad)
  {
    _eigenstrain[_qp] = _eigen_ortho_tensor * (_temperature[_qp].value() - _stress_free_temperature[_qp]);
  }
  else
  {
    _eigenstrain[_qp] = _eigen_ortho_tensor * (_temperature[_qp].value() - _stress_free_temperature[_qp]);
    
    (*_deigenstrain_dT)[_qp].zero();
    if (!_use_old_temperature)
      (*_deigenstrain_dT)[_qp] = _eigen_ortho_tensor;
  }

}


template class ComputeThermalExpansionEigenstrainOrthotropicTempl<false>;
template class ComputeThermalExpansionEigenstrainOrthotropicTempl<true>;

```
Does this make sense ? Or is it a convoluted way of doing such ? Would be great to have some feedback

Thanks,
-Aritra

**Reply by chakra34** [reply, marked as answer]
I tested the above code and it seems to work for a preliminary case, where I have CTE_11 != CTE_22 != CTE_33 for a cubic geometry with linearly varying temperature field:

![image](https://github.com/idaholab/moose/assets/22983992/5afdcef8-d5fe-49ff-8bca-e28def3d2cd8)

**Reply by chakra34** [reply]
However, since I inherited from ```ComputeThermalExpansionEigenstrain``` which needs the required parameter ```thermal_expansion_coeff``` I had to provide that in the input file (i gave it 0).

**Reply by wangzhaohao** [reply]
I guess define the same name
'thermal_expansion_coeff' will slove this problem？Later i will try.
